languages/javalang/builder-renderer/commercetools-java-client/src/main/java/client/ApiHttpClient.java

```diff
@@ -1,9 +1,45 @@
 package client;
 
-import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 
-public interface ApiHttpClient {
+public class ApiHttpClient {
 
-    public ApiHttpResponse execute(final ApiHttpRequest httpRequest) throws IOException;
-    
+    private final Middleware reducedMiddleware;
+
+    public ApiHttpClient(final List<Middleware> middlewares) {
+        final List<Middleware> middlewareList = new ArrayList<>();
+        middlewareList.addAll(middlewares);
+        middlewareList.add(NoOpMiddleware.getInstance());
+        this.reducedMiddleware = middlewareList.stream().reduce(ApiHttpClient::reduceMiddlewares).get();
+    }
+
+    public CompletableFuture<ApiHttpResponse> execute(final ApiHttpRequest request) {
+        return reducedMiddleware
+                .next(MiddlewareArg.from(request, null, null, null))
+                .thenApply(middlewareArg -> {
+                    if (middlewareArg.getError() == null) {
+                        return middlewareArg.getResponse();
+                    }
+                    throw new CompletionException(middlewareArg.getError());
+                });
+    }
+
+    public ApiHttpResponse executeBlocking(ApiHttpRequest request) {
+        try {
+            return execute(request).get();
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
Here we should use proper logging or remove the code.

Reply:
actually the logging should be totally absent from here, it should be handled at the logger middleware.